// miso.dataset/require_example/app/config.js
// Set the require.js configuration for your application.
require.config({

  // Initialize the application with the main application file.
  deps: ["main"],

  paths: {
    // JavaScript folders.
    libs: "../assets/js/libs",
    plugins: "../assets/js/plugins",

    // Libraries.
    jquery: "../assets/js/libs/jquery",
    lodash: "../assets/js/libs/lodash",
    moment: "../assets/js/libs/moment",
    'underscore.deferred': "../assets/js/libs/underscore.deferred",
    'underscore.math': "../assets/js/libs/underscore.math",
    backbone: "../assets/js/libs/backbone",
    'miso.ds': "../assets/js/libs/miso.ds.0.4.0.m"
  },

  shim: {
    
    // Backbone library depends on lodash and jQuery.
    backbone: {
      deps: ["lodash", "jquery"],
      exports: "Backbone"
    },

    'underscore.deferred' : {
      deps : ["lodash"],
      exports : "_"
    },

    'underscore.math' : {
      deps : ["lodash"],
      exports : "_"
    },

    'miso.ds' : {
      deps: ["lodash", "moment", "underscore.deferred", "underscore.math"],
      exports: "Miso"
    },

    // Backbone.LayoutManager depends on Backbone.
    "plugins/backbone.layoutmanager": ["backbone"]
  }

});
